### PR TITLE
Thread pool

### PR DESCRIPTION
**Pull request by charlesgregory**
This PR adds wrappers around `htsThreadPool` and `hts_tpool`. The goal would be to allow parallel, in-order execution of arbitrary functions. These jobs could be added directly to the same thread pools that are used for reading and writing files.

### CODE REVIEW

**Comment by jblachly** [comment]
What is the use case? I may not be qualified to comment on this 🤯

File: source/dhtslib/file/iterator.d [hunk 1]
```diff
@@ -142,7 +142,7 @@ if(is(T == Bam1) || is(T == Bcf1) || is(T == Kstring))
     /// If you keep the result around it should be duplicated
     T front()
     {
-        return rec;
+        return rec.dup;
```
Comment (jblachly) [inline source/dhtslib/file/iterator.d:145]: Where is the record (result) freed?

Comment (jblachly) [reply source/dhtslib/file/iterator.d:145]: Working with C libraries is hard

File: source/dhtslib/threadpool.d [hunk 2]
```diff
@@ -0,0 +1,482 @@
+/**
+This module provides multithreading support through htslib. It allows for mapping delagates,
+and functions via a range interface with `parallelMap`. Crucially, the htslib task pool performs
```
Comment (jblachly) [inline source/dhtslib/threadpool.d:3]: Sounds amazing, but can you explain the concrete use case(s)?

For example, doing some sort of variant calling or other metric collection simply by iterating over a BAM file perhaps?

File: source/dhtslib/threadpool.d [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+{
+    enum bool noUnsharedAliasing = !hasUnsharedAliasing!T;
+}
+
+// This template tests whether a function may be executed in parallel from
+// @safe code via Task.executeInNewThread().  There is an additional
+// requirement for executing it via a TaskPool.  (See isSafeReturn).
+private template isSafeTask(F)
+{
+    enum bool isSafeTask =
```
Comment (jblachly) [inline source/dhtslib/threadpool.d:42]: I've not templated an enum before. How does this work? In particular, if you are declaring `enum bool isSafeTask =`, when you instantiate the template like `isSafeTask!foo` does that yield a bool? I find that weird because it's not a function returning anything. I guess I need to review non-function templates.

File: source/dhtslib/threadpool.d [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+        (functionAttributes!F & (FunctionAttribute.safe | FunctionAttribute.trusted)) != 0 &&
+        (functionAttributes!F & FunctionAttribute.ref_) == 0 &&
+        (isFunctionPointer!F || !hasUnsharedAliasing!F) &&
+        allSatisfy!(noUnsharedAliasing, Parameters!F);
+}
+
+/// Thread pool instance. Wraps htsThreadPool and hts_tpool. 
+struct ThreadPool
```
Comment (jblachly) [inline source/dhtslib/threadpool.d:50]: Is -1 an htslib magic constant, or ours? (I think hts lib, but i quickly searched in thread_pool.h and couldn't find it)

Should document for this struct any magic values irrespective

File: source/dhtslib/threadpool.d [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+    @property tpool() {
+        return this.htsPool.pool;
+    }
+
+}
+
+
+/**
+Below code is ported from std.parallelism to work with htslib's thread pool.
+It is BSL1.0 licensed.
```
Comment (jblachly) [inline source/dhtslib/threadpool.d:89]: This looks pretty MIT like. I think to be safe you should include the license notice as a file in the root of the Repo now, like LICENSE.BSL1.0 or something

File: source/dhtslib/threadpool.d [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+
+    @property tpool() {
+        return this.htsPool.pool;
+    }
+
+}
+
+
+/**
+Below code is ported from std.parallelism to work with htslib's thread pool.
```
Comment (jblachly) [inline source/dhtslib/threadpool.d:88]: I'm explicitly not reviewing any of the code directly adapted from std.parallelism

File: source/dhtslib/threadpool.d [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+    
+    this(R range, ThreadPool * pool) {
+        this.pool = pool;
+        this.range = range;
+        this.proc = HtsProcess(hts_tpool_process_init(this.pool.tpool, cast(int)this.pool.threads * 2, 0));
+        frontVal = frontVal.init;
+        auto qsize =  hts_tpool_process_qsize(this.proc);
+        for(auto i = 0; i < qsize; i++) {
+            if(this.range.empty) {
+                break;
```
Comment (jblachly) [inline source/dhtslib/threadpool.d:420]: empty range condition seems important to test -- why do yoyu think it was not covered?

File: source/htslib/bgzf.d [hunk 8]
```diff
@@ -34,6 +34,7 @@ import core.sys.posix.sys.types;
 
 import htslib.hfile : hFILE;
 import htslib.kstring;
+import htslib.thread_pool;
```
Comment (jblachly) [inline source/htslib/bgzf.d]: Is it actually used in this file?

File: source/htslib/kstring.d [hunk 9]
```diff
@@ -199,6 +199,13 @@ void ks_free(kstring_t* s)
     }
 }
 
+kstring_t* ks_dup(kstring_t* s) {
```
Comment (jblachly) [inline source/htslib/kstring.d:202]: Interestingly is not a part of kstring.h or kstring.c ; will this make it more difficult to keep your file in sync with upstream?

File: source/dhtslib/file/iterator.d (hunk 1 above)
Comment (charlesgregory) [reply source/dhtslib/file/iterator.d:145]: Rec here is a `SafeHtslibPtr` instance so it is reference counted. The `dup` function is just a wrapper around the particular htslib dup function used for the type wrapped by `SafeHtslibPtr`. So for a `bam1_t` it is `SafeHtslibPtr!(bam1_t, bam_destroy1, bam_dup1);`.

File: source/dhtslib/threadpool.d (hunk 2 above)
Comment (charlesgregory) [reply source/dhtslib/threadpool.d:3]: So my intended use is primarily to be able to do work in parallel in an in-order fashion. This can be done via `std.parallelism` though I have never been able to get it to work for me. Though this also can provide the advantage of sharing the thread pool with readers and writers. Then whatever jobs are taking the most work will be taking a majority of the threads. So if i have 32 cores to deal with, if my work to be done is more compute intensive than the need to decompress and do file I/O, it can use a majority of those threads.

So my intended use case would be something like the following:
```d
// set up global pool
enforceGlobalPoolThreads(32);

// use global pool for multi-threaded decompression
auto vcfr = VcfReader("a.vcf.gz");
vcfr.setThreadPool(globalPool);

// use global pool for multi-threaded compression/writing
auto vcfw = VcfWriter("b.vcf.gz", vcfr.header);
vcfw.setThreadPool(globalPool);
vcfw.writeHeader();

// perform work in parallel but in order
auto range = vcfr.parallelMap((rec) {
    // Even the reading of the gff3 could use the same thread pool
    rec.filterAnnnotationsToPrincipalIsoform(rec.infos["ANN"], "gencode.gff3.gz");
    return rec;
})

// Though this loop is single threaded, the range it is pulling from is not
// And the writer is writing the results in parallel
// Since data is returned in order, no need to re-sort later
foreach(rec;range) {
    vcfw.write(rec);
}
```

In a similar vein, I could use this in `fade` for preventing the need to resort bam files after artifact identification before filtering artifact reads, while retaining the ability to do work in parallel.

File: source/dhtslib/threadpool.d (hunk 3 above)
Comment (charlesgregory) [reply source/dhtslib/threadpool.d:42]: It's just another form of eponymous templates. It is used as compile-time logic like most of the functions in `std.traits`. So I could evaluate at compile-time the task to be executed is `@safe`. So this would mainly be used in like a `static if` statement.

File: source/dhtslib/threadpool.d (hunk 4 above)
Comment (charlesgregory) [reply source/dhtslib/threadpool.d:50]: No it is simply my default value to indicate "set threads to number of cpus". Will add documentation.

File: source/dhtslib/threadpool.d (hunk 5 above)
Comment (charlesgregory) [reply source/dhtslib/threadpool.d:89]: Will do.

File: source/dhtslib/threadpool.d (hunk 7 above)
Comment (charlesgregory) [reply source/dhtslib/threadpool.d:420]: I simply did not add a test for an empty range. Will add.

File: source/htslib/bgzf.d (hunk 8 above)
Comment (charlesgregory) [reply source/htslib/bgzf.d]: Needed for the opaque `hts_tpool` struct declaration. Before I would have to cast a `hts_tpool` from `htslib.bgzf` to the `hts_tpool` type in `htslib.thread_pool`, if I needed to use the same thread pool for both `htslib.bgzf` and other `htslib` functions.

Comment (charlesgregory) [reply source/htslib/bgzf.d]: I should probably limit the import to only `hts_tpool`.

File: source/htslib/kstring.d (hunk 9 above)
Comment (charlesgregory) [reply source/htslib/kstring.d:202]: Since we have the diff system, probably not. Though I should update the diffs considering I have new changes that need to carry forward. 

Though I also thought that `kstring` already had a `dup` function and was surprised to find that it didn't.

File: source/dhtslib/threadpool.d (hunk 2 above)
Comment (jblachly) [reply source/dhtslib/threadpool.d:3]: Makes sense. When you say have never been able to get std parallelism to work for you, what does that mean with respect to the fact that you have imported its code? Do you expect it to work? S is it tested already and confirmed working?

Comment (charlesgregory) [reply source/dhtslib/threadpool.d:3]: I have never been able to get `std.parallelism` 's provided functions to work for my purposes. `std.parallelism`'s `TaskPool.map` does in-order but evaluates lazily, and in my testing didn't work very well. `std.parallelism`'s `TaskPool.amap` is similar but an eager algorithm but requires a `RandomAccessRange` as input. `std.parallelism`'s `parallel foreach` function works pretty well but doesn't output a range and is not in-order. 

The code I have ported from `std.parallelism` works as intended. It eagerly evaluates tasks in-order and is more leinient to which delegates it can run. Though it is likely not quite as safe, due to some more manual memory management. 

Though the user shouldn't have to worry about it. Notably, the code I wrote is actually executed serially,  only the tasks are executed in parallel via htslib and htslib's `hts_tpool_next_result_wait` makes it easy to synchronously grab the next ready result as it becomes available. Making wrapping it in a range pretty easy.

Comment (charlesgregory) [reply source/dhtslib/threadpool.d:3]: Additionally, my `parallelMap` function can use a simple `InputRange` as input.

File: source/dhtslib/threadpool.d (hunk 6 above)
Comment (charlesgregory) [reply source/dhtslib/threadpool.d:88]: Most of `Task` is ripped directly because I needed their compile time logic to make it work with any function/delegate. I also tried to port `parallel` that could be used in a `foreach` statement, but it was too annoying and I don't really need it anyway.
